api-server/scanner-lib/src/blockchain_state/mod.rs

```diff
@@ -159,7 +137,393 @@ impl<S: ApiServerStorage + Send + Sync> LocalBlockchainState for BlockchainState
+                    }
+                }
+            }
+            TxInput::Account(outpoint) => match outpoint.account() {
+                AccountSpending::DelegationBalance(_, withdraw_amount) => {
+                    (CoinOrTokenId::Coin, *withdraw_amount)
+                }
+            },
+            TxInput::AccountCommand(_, account_op) => match account_op {
+                AccountCommand::MintTokens(token_id, amount) => {
+                    (CoinOrTokenId::TokenId(*token_id), *amount)
+                }
+                AccountCommand::UnmintTokens(_)
+                | AccountCommand::LockTokenSupply(_)
+                | AccountCommand::FreezeToken(_, _)
+                | AccountCommand::UnfreezeToken(_)
+                | AccountCommand::ChangeTokenAuthority(_, _) => {
+                    continue;
+                }
+            },
+        };
+
+        grouped
+            .entry(coin_or_token_id)
+            .and_modify(|a: &mut Amount| *a = (*a + amount).expect("should not overflow"))
+            .or_insert(amount);
+    }
+
+    Ok(grouped)
+}
+
+async fn calculate_total_outputs<T: ApiServerStorageRead>(
+    db_tx: &mut T,
+    outputs: &[TxOutput],
+) -> Result<BTreeMap<CoinOrTokenId, Amount>, ApiServerStorageError> {
+    let mut grouped = BTreeMap::new();
+
+    for output in outputs {
+        let output_value = match output {
+            TxOutput::Transfer(v, _) | TxOutput::LockThenTransfer(v, _, _) | TxOutput::Burn(v) => {
+                v.clone()
+            }
+            TxOutput::CreateStakePool(_, data) => OutputValue::Coin(data.value()),
+            TxOutput::ProduceBlockFromStake(_, pool_id) => {
+                let pledge_amount = db_tx
+                    .get_pool_data(*pool_id)
+                    .await?
+                    .expect("pool should be present")
+                    .pledge_amount();
+                OutputValue::Coin(pledge_amount)
+            }
+            TxOutput::DelegateStaking(v, _) => OutputValue::Coin(*v),
+            TxOutput::CreateDelegationId(_, _)
+            // IssueFungibleToken cannot be spent so the output value is 0
+            | TxOutput::IssueFungibleToken(_)
+            // when in tx output IssueNft has 0 output value because it is created there
+            | TxOutput::IssueNft(_, _, _)
+            | TxOutput::DataDeposit(_) => continue,
+        };
+        let (coin_or_token_id, amount) = match output_value {
+            OutputValue::Coin(amount) => (CoinOrTokenId::Coin, amount),
+            OutputValue::TokenV1(token_id, amount) => (CoinOrTokenId::TokenId(token_id), amount),
+            OutputValue::TokenV0(_) => {
+                continue;
+            }
+        };
+
+        grouped
+            .entry(coin_or_token_id)
+            .and_modify(|a: &mut Amount| *a = (*a + amount).expect("should not overflow"))
+            .or_insert(amount);
+    }
+
+    Ok(grouped)
+}
+
+async fn calculate_fees<T: ApiServerStorageWrite>(
```

Comment:
We should be using the accumulator for this. Repeating this and not using the functions that we have in the tx verifier is a big risk as it can conflict with what consensus does in the future if it changes.

Reply:
fixed